Support LLVM 12.

dan-zheng opened the pull request:
Update `llvm-hs` from LLVM 9 to LLVM 12 (current top-of-tree).

List of changes:

```
- Attributes
  - Various minor changes, not listed in detail.
- Debug information
  - DIFlag: BlockByrefStruct and ArgumentNotModified removed.
  - DISubrange
    - New arguments: upperBound and stride.
    - Arguments {lowerBound,upperBound,stride} are generalized and may be
      DIVariable or DIExpression in addition to Constant Int64.
  - DIModule
    - New arguments: apiNotesFile and lineNo.
    - Removed arguments: isysroot.
- Instructions
  - New instructions: freeze.
  - Modified instructions
    - shufflevector: now takes mask as [Int32] instead of Constant.
- JIT
  - Some deprecated OrcJIT v1 APIs are now removed: VModuleKey, SymbolResolver.
  - Start updating APIs and tests using OrcJIT v2 APIs. Work-in-progress.
- Target
  - Target option flag changes, not listed in detail.
- Transforms
  - Some existing transforms were removed or merged into others.
    Some had options modified.

- Newly added APIs
  - Added Triple (target triple) API to llvm-hs-pure, along with string
    conversion and parsing.
  - Added debugging utilities: dumpMetadata.
- Tests
  - Textual LLVM IR changes
    - Function parameters are now printed with local identifiers (e.g. `%0`).
    - Instruction alignment arguments must now be explicit: 0 no longer seems
      like a valid default value.
  - Instrumentation tests
    - Update MemorySanitizer test to run only on supported platforms: matching
      the implementation at https://github.com/llvm/llvm-project/blob/release/12.x/llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp.
  - Optimization tests
    - Add a more robust "is module vectorized?" check for the SLPVectorization
      test.
```

Current status: all tests pass via `stack test`.

Todos:
- [x] Relevant comments to resolve: `{FIXME,TODO,NOTE}(llvm-12)`
- [x] Lots of code cleanup and API design finalization for OrcJIT v2.
- [x] Update installation instructions for `llvm-hs@llvm-12`.
- [x] Fix CI for `llvm-hs@llvm-12`.

apaszke commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,72 @@
+# TODO(llvm-12): Finish implementing this CI workflow.
```
(As discussed offline) Let's keep the CI changes as a follow-up to make this PR smaller.

apaszke commented on llvm-hs-pure/src/LLVM/Triple.hs:
```diff
@@ -0,0 +1,380 @@
+{-# LANGUAGE RecordWildCards #-}
```
(As discussed offline) This is nice to have, but not crucial for this PR. Let's split it out into a follow-up change.

apaszke commented on llvm-hs/llvm-hs.cabal:
```diff
@@ -129,6 +129,7 @@ library
     LLVM.Internal.Module
     LLVM.Internal.ObjectFile
     LLVM.Internal.OrcJIT
+    LLVM.Internal.OrcJITV2
```
OrcJIT v1 has been deprecated for quite a while now, so I would lean towards removing the old bindings and renaming OrcJITV2 -> OrcJIT. I wouldn't be surprised if the release version of LLVM 12 would not include OrcJIT v1 bindings.

apaszke commented on llvm-hs/src/LLVM/Internal/FFI/BitcodeC.cpp:
```diff
@@ -16,7 +16,7 @@ LLVMModuleRef LLVM_Hs_ParseBitcode(
 	char **error
 ) {
-    Expected<std::unique_ptr<Module>> moduleOrErr = parseBitcodeFile(unwrap(mb)->getMemBufferRef(), *unwrap(c));
+    Expected<std::unique_ptr<Module>> moduleOrErr = parseBitcodeFile(unwrap(mb)->getMemBufferRef(), *unwrap(c), [](StringRef) { return None; });
```
`parseBitcodeFile` has a default argument for this callback, which is equivalent to what you're passing in here. I think that the old code should work here.

apaszke commented on llvm-hs/src/LLVM/Internal/FFI/BuilderC.cpp:
```diff
@@ -280,6 +280,10 @@ LLVMValueRef LLVM_Hs_BuildInBoundsGEP(LLVMBuilderRef B, LLVMValueRef Pointer,
 }
 
+LLVMValueRef LLVM_Hs_Freeze(LLVMBuilderRef B, LLVMValueRef Op, const char *Name) {
+    return wrap(unwrap(B)->Insert(new FreezeInst(unwrap(Op), Name)));
```
Looks like [this is already part of the C API](https://github.com/llvm/llvm-project/blob/8bd8534aa3bdddf328683d79cfa46ee1d678f3d2/llvm/include/llvm-c/Core.h#L3959), so we shouldn't be reimplementing it.

apaszke commented on llvm-hs-pure/src/LLVM/AST/Operand.hs:
```diff
@@ -131,7 +131,7 @@ data DIFlag
   | FwdDecl
   | AppleBlock
-  | BlockByrefStruct
+  | ReservedBit4 -- Used to be BlockByRef, can be reused for anything except DICompositeType.
```
I understand that this is useful, because the ADT positions are meaningful for some Template Haskell code generation later? Otherwise we should remove it.

apaszke commented on llvm-hs/src/LLVM/Internal/FFI/InstructionC.cpp:
```diff
@@ -136,7 +137,7 @@ unsigned LLVM_Hs_GetInstrAlignment(LLVMValueRef l) {
 void LLVM_Hs_SetInstrAlignment(LLVMValueRef l, unsigned a) {
 	switch(unwrap<Instruction>(l)->getOpcode()) {
-#define ENUM_CASE(n) case Instruction::n: unwrap<n ## Inst>(l)->setAlignment(a); break;
+#define ENUM_CASE(n) case Instruction::n: unwrap<n ## Inst>(l)->setAlignment(MaybeAlign(a).valueOrOne()); break;
```
Can't you construct an instance of `Align` directly here? There is no need for the `Maybe` part since you always have `a`.

apaszke commented on llvm-hs/src/LLVM/Internal/FFI/InstructionC.cpp:
```diff
@@ -266,6 +267,15 @@ unsigned LLVM_Hs_GetMetadata(
+
+void LLVM_Hs_GetShuffleVectorMask(LLVMValueRef i, unsigned *result) {
+  auto mask = unwrap<ShuffleVectorInst>(i)->getShuffleMask();
+	std::copy(mask.begin(), mask.end(), result);
```
nit: weird indentation

apaszke replied:
Also, for functions that write to a pointer like that, it is often good to pass in the length of the pointer, just so that you can check that this function will never write out-of-bounds (with respect to the caller's expectations).

apaszke commented on llvm-hs/src/LLVM/Internal/FFI/OrcJIT.hs:
```diff
@@ -70,3 +105,4 @@ foreign import ccall "wrapper" wrapGetSymbolResolver ::
 
 foreign import ccall "wrapper" wrapSetSymbolResolver ::
   (ModuleKey -> Ptr SymbolResolver -> IO ()) -> IO (FunPtr (ModuleKey -> Ptr SymbolResolver -> IO ()))
+-}
```
Please remove this comment. It's hard to review this file with it, because git can't figure out the right diff.

apaszke commented on llvm-hs/src/LLVM/Internal/FFI/OrcJIT/CompileLayer.hs:
```diff
@@ -26,9 +28,13 @@ foreign import ccall safe "LLVM_Hs_CompileLayer_addModule" addModule ::
+{-
 foreign import ccall safe "LLVM_Hs_CompileLayer_findSymbolIn" findSymbolIn ::
   Ptr CompileLayer -> ModuleKey -> CString -> LLVMBool -> IO (Ptr JITSymbol)
+-}
```
Yes, let's remove them

apaszke commented on llvm-hs/src/LLVM/Internal/FFI/OrcJITC.cpp:
```diff
@@ -2,6 +2,12 @@
 #pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 #pragma clang diagnostic ignored "-Wdeprecated-declarations"
 
+// FIXME(llvm-12): Clean up this file.
```
Let's just remove this file for simplicity.

apaszke commented on llvm-hs/src/LLVM/Internal/FFI/PassManagerC.cpp:
```diff
@@ -124,17 +120,19 @@ void LLVM_Hs_AddGCOVProfilerPass(
-	options.FunctionNamesInData = functionNamesInData;
+	options.Atomic = atomic;
+	options.Filter = filter;
+  options.Exclude = exclude;
```
nit: weird indentation

apaszke commented on llvm-hs/src/LLVM/Internal/FFI/PassManagerC.cpp:
```diff
@@ -169,6 +168,29 @@ void LLVM_Hs_AddBoundsCheckingPass(LLVMPassManagerRef PM) {
+/*
+void LLVM_Hs_AddConstantPropagationPass(LLVMPassManagerRef PM) {
+	unwrap(PM)->add(createConstantPropagationPass());
+}
```
Removed in https://reviews.llvm.org/D85159

apaszke commented on llvm-hs/src/LLVM/Internal/FFI/PassManagerC.cpp:
```diff
@@ -169,6 +168,29 @@ void LLVM_Hs_AddBoundsCheckingPass(LLVMPassManagerRef PM) {
+	unwrap(PM)->add(createDeadInstEliminationPass());
+}
+
+void LLVM_Hs_AddIPConstantPropagationPass(LLVMPassManagerRef PM) {
```
For reference: https://github.com/llvm/llvm-project/commit/4ffa6a27aca17fe88fa6bdd605b198df6632a570

apaszke commented on llvm-hs/src/LLVM/Internal/FFI/PassManagerC.cpp:
```diff
@@ -169,6 +168,29 @@ void LLVM_Hs_AddBoundsCheckingPass(LLVMPassManagerRef PM) {
 	unwrap(PM)->add(createBoundsCheckingLegacyPass());
 }
 
+// TODO(llvm-12): Confirm that these passes have been removed in LLVM 9 → LLVM 12.
```
I looked through them and I think they're good to remove, so let's get rid of them

apaszke commented on llvm-hs/src/LLVM/Internal/OrcJIT/IRCompileLayer.hs:
```diff
@@ -46,7 +46,9 @@ newIRCompileLayer linkingLayer (TargetMachine tm) = flip runAnyContT return $ do
 withIRCompileLayer :: LinkingLayer l => l -> TargetMachine -> (IRCompileLayer l -> IO a) -> IO a
 withIRCompileLayer linkingLayer tm =
   bracket (newIRCompileLayer linkingLayer tm) disposeCompileLayer
+-}
```
??

apaszke commented on llvm-hs/src/LLVM/OrcJIT/CompileLayer.hs:
```diff
@@ -2,10 +2,11 @@ module LLVM.OrcJIT.CompileLayer
+  -- , findSymbolIn
+  -- , addModule
+  -- , removeModule
+  -- , withModule
```
The fact that they're unused in this project doesn't mean that no one is using them outside of the library. Since this is not part of the `Internal` namespace, we should do our best to preserve them.

apaszke commented on llvm-hs/src/LLVM/Transforms.hs:
```diff
@@ -17,16 +17,15 @@ data Pass
   | GlobalValueNumbering { noLoads :: Bool }
   | InductionVariableSimplify
   | InstructionCombining
+  | InstructionSimplify
```
Please add a comment saying that this does constant propagation, for better discoverability.

apaszke commented on llvm-hs/test/LLVM/Test/Instrumentation.hs:
```diff
@@ -33,6 +35,8 @@ import qualified LLVM.AST.Attribute as A
 import qualified LLVM.AST.Global as G
 import qualified LLVM.AST.Constant as C
 
+import Debug.Trace
```
Please remove this

apaszke commented on llvm-hs/test/LLVM/Test/Instrumentation.hs:
```diff
@@ -144,23 +148,41 @@ ast = do
-     ("MemorySanitizer", defaultMemorySanitizer),
-     ("ThreadSanitizer", defaultThreadSanitizer),
-     ("BoundsChecking", BoundsChecking)--,
+      shouldTestPass' <- shouldTestPass
```
nit: perhaps rename this to `runTest <- shouldRunTest`

apaszke commented on llvm-hs/test/LLVM/Test/Instrumentation.hs:
```diff
@@ -144,23 +148,41 @@ ast = do
+        else do
+          cpu <- getHostCPUName
+          triple <- getProcessTargetTriple
+          bool <- isMemorySanitizerSupported
```
This probably needs a cleanup

apaszke commented on llvm-hs/test/LLVM/Test/Optimization.hs:
```diff
@@ -29,11 +30,22 @@ import qualified LLVM.AST.CallingConvention as CC
+dumpModule' :: A.Module -> IO ()
+dumpModule' m = withContext $ \context -> withModuleFromAST context m $ \m' -> do
+  mPtr <- M.readModule m'
+  M.dumpModule mPtr
```
nit: This whole lambda (starting from `\m'`) is just `M.readModule >>= M.dumpModule`

apaszke commented on llvm-hs/test/LLVM/Test/ParameterAttribute.hs:
```diff
@@ -31,7 +31,7 @@ instance Arbitrary ParameterAttribute where
     , return InReg
     , return SRet
-    , Alignment <$> elements (map (2^) [0..30 :: Int])
+    , Alignment <$> elements (map (2^) [0..29 :: Int])
```
What's wrong with 30?

apaszke commented on llvm-hs/test/LLVM/Test/OrcJIT.hs:
```diff
@@ -159,5 +176,27 @@ tests =
+              -- FIXME(llvm-12): "main" vs "_main" symbol name seems platform-dependent,
+              -- to be verified. "main" on Linux and "_main" on macOS. Find a
+              -- robust platform-independent fix – perhaps by reviving
+              -- `OrcV2.findSymbolIn` which takes a `MangledSymbol`.
```
Yup, we should revive `MangledSymbol` for OrcJITV2

apaszke commented on llvm-hs/test/LLVM/Test/OrcJIT.hs:
```diff
@@ -159,5 +176,27 @@ tests =
+            OrcV2.withRTDyldObjectLinkingLayer es $ \ol ->
+            OrcV2.withIRCompileLayer es ol tm $ \il -> do
+              dl <- getTargetMachineDataLayout tm
+              dylib' <- OrcV2.getJITDylibByName es dylibName
```
Is this any different than `dylib`?

apaszke commented on llvm-hs/test/LLVM/Test/OrcJIT.hs:
```diff
@@ -76,6 +89,10 @@ moduleTransform passmanagerSuccessful modulePtr = do
 tests :: TestTree
 tests =
   testGroup "OrcJit" [
+    -- FIXME(llvm-12): Re-enable tests.
```
Just leaving a comment so that we don't forget to resolve this before we're done

apaszke commented on llvm-hs/src/LLVM/OrcJIT.hs:
```diff
@@ -3,9 +3,10 @@ module LLVM.OrcJIT (
-    addModule,
-    removeModule,
-    withModule,
+    -- TODO(llvm-12): Remove unused APIs.
```
As mentioned elsewhere: we should preserve as much of the public API as we can. It's a library, so you can't assume there are no users

apaszke commented on llvm-hs/src/LLVM/Internal/FFI/OrcJITV2C.cpp:
```diff
@@ -0,0 +1,162 @@
+    }
+}
+
+}
```
This needs a lot of cleanups, but you know this

apaszke reviewed (commented):
This mostly LGTM, but it still requires some polish. The ownership model for modules around OrcJIT is a little janky, but I guess it's fine for now. I would suggest merging with the cloning at `ThreadSafeModule` construction time, and later we can introduce the double-indirection we've been talking about to remove the clone (and still have the bracketing APIs do the right thing).

sureyeaah committed:
Fix Optimization.hs test: willreturn attributes are now inferred.

https://github.com/llvm/llvm-project/commit/65fd034b95d69fa0e634861ee165b502ceb92a12

apaszke committed:
Remove OrcJITv1, lots of cleanups and fixes.

apaszke committed:
Restore getSymbolAddressInProcess

andrew-wja commented:
Hi Dan, could you update this PR to target the `llvm-12-unstable` branch? You may have to do a rebase, since some fixes have been applied to the `llvm-9` branch and it would be great to have them in this PR without cherry-picking each commit over later. Thanks!

dan-zheng commented:
Sure thing, done!
